### ISSUE TITLE
Closing asterisk */ not on its own line not detected for google style

### ISSUE THREAD

**Issue by Praveen7294** [checkstyle/checkstyle#18273]
From: [7.1.1 General Form](https://google.github.io/styleguide/javaguide.html#s7.1.1-javadoc-multi-line) and https://github.com/google/google-java-format/issues/1279#issuecomment-3320778635

7.1.1 General form
The basic formatting of Javadoc blocks is as seen in this example:

```
/**
 * Multiple lines of Javadoc text are written here,
 * wrapped normally...
 */
public int method(String p1) { ... }
```
... or in this single-line example:
`/** An especially short bit of Javadoc. */`
The basic form is always acceptable. The single-line form may be substituted when the entirety of the Javadoc block (including comment markers) can fit on a single line.

---
Part of [17778](https://github.com/checkstyle/checkstyle/issues/17778), subproblem 3

According to Google Java Style, the preferred form is to place `/**` and `*/` on their own lines unless the entire Javadoc comment fits on a single line. However, in `google_checks.xml`, the [JavadocLeadingAsteriskAlign](https://checkstyle.org/checks/javadoc/javadocleadingasteriskalign.html) check validates the alignment of the closing asterisk only when whitespace appears before the `*/`. It ignore the alignment of leading asterisk when non-whitespace characters appear before the `*/`. The following example is:
```
  /**
   * Correct Indentation for leading asterisk. */
  private void foo2() {}
```
If we can add support for validating the alignment of the closing `*/`, it would improve compliance with the Google Javadoc style.

#### Google Formatter

Test.java
```
public class Test {
  
  /**
   * void constructor.
   *
   * @param a testing... */
  public Test(int a) {}

}
```
cli:
`$ java -jar google-java-format-1.33.0-all-deps.jar Test.java > FormattedCode.java`

Formatted Code:
```
public class Test {

  /**
   * void constructor.
   *
   * @param a testing...
   */
  public Test(int a) {}
}
```
Cli:
```
$ diff -Naru Test.java FormattedCode.java
--- Test.java   2025-12-12 00:50:02.874673400 +0530
+++ FormattedCode.java  2025-12-12 00:51:12.628082900 +0530
@@ -1,9 +1,9 @@
 public class Test {
-
+
   /**
    * void constructor.
    *
-   * @param a testing... */
+   * @param a testing...
+   */
   public Test(int a) {}
-
 }
```

**Comment by Zopsss** [comment]
> It does not validate cases where non-whitespace characters appear before the `*/`, even though this situation is explicitly described in the check’s documentation. As a result, the following example is not flagged:

@Praveen7294 The check has explicitly mentioned that if the endling javadoc line contains a leading asterisk then that asterisk's alignment will be considered:

```
If the ending javadoc line contains a leading asterisk, then that leading asterisk's alignment will be considered, the closing Javadoc tag will be ignored.
```

The check was designed to handle only leading asterisk's alignments, it does not cover such edge cases.

**Comment by Zopsss** [comment]
Please add: 

```
Part of [17778](https://github.com/checkstyle/checkstyle/issues/17778), subproblem 3
```

to keep track of issue clearly.

**Comment by Praveen7294** [comment]
> [@Praveen7294](https://github.com/Praveen7294) The check has explicitly mentioned that if the endling javadoc line contains a leading asterisk then that asterisk's alignment will be considered:
> 
> ```
> If the ending javadoc line contains a leading asterisk, then that leading asterisk's alignment will be considered, the closing Javadoc tag will be ignored.
> ```
> 
> The check was designed to handle only leading asterisk's alignments, it does not cover such edge cases.

you are right, i have corrected my statement.

**Comment by Zopsss** [comment]
> If we can add support for validating the alignment of the closing `*/`, it would improve compliance with the Google Javadoc style.

Yes, we can implement this. Currently, the Check ignores the closing `*/` when another leading asterisk appears on the same line. However, we can introduce a new property that enforces indentation validation for `*/` regardless of whether a leading asterisk is present.

@romani What do you think?

Just to clarify, when I originally implemented this Check, it was designed as a general-purpose solution rather than specifically targeting Google Style. Since some users may not want or expect this stricter behavior, it would be best to add this as an optional, configurable property that they can enable as needed.

**Comment by romani** [comment]
`*/` has no leading asterisk.
We just need new Check that alight `/*` and `*/` (comment start and comment end)
https://docs.oracle.com/javase/specs/jls/se11/html/jls-3.html#jls-3.7

We need new check for this, as users like to do alignments in different ways.
see example at https://docs.scala-lang.org/style/scaladoc.html , Quote from it:
Scaladoc style, with gutter asterisks aligned in column two:
```
/** Provides a service as described.
 *
 *  This is further documentation of what we're documenting.
 *  Here are more details about how it works and what it does.
 */
def member: Unit = ()
```

Scaladoc style, with gutter asterisks aligned in column three:
```
/** Provides a service as described.
  *
  * This is further documentation of what we're documenting.
  * Here are more details about how it works and what it does.
  */
def member: Unit = ()
```

I do not recommend to ad support for alignment of third position from the beginning, but as we have new Check , extension of it will be easy.

Lets create new issue and start discussion of new Check. Ones implementation is done, we can comeback to this issue and activate it google style config.

**Comment by Praveen7294** [comment]
@romani 
- Indentation of `/**` (Opening javadoc tag) is checked by `CommentsIndentation`.
- Alignment of `*` (leading asterisk) is checked by `JavadocLeadingAsteriskAlign` with respect to `/**`. Therefore, if the position of `*` is incorrect, a violation will also be reported for the leading asterisk alignment.
- Alignment of `*/` is checked by `JavadocLeadingAsteriskAlign` when it has it's own line.

The only case not checked by any existing check is when `*/` does not have its own line. Therefore, introducing a new property only for this case is a good idea.

> However, we can introduce a new property that enforces indentation validation for */ regardless of whether a leading asterisk is present.

@Zopsss i'm ok with this.

**Comment by romani** [comment]
last `*` before `/` (together `*/` ) is not a leading asterisk, so we can not reuse JavadocLeadingAsteriskAlign for rules for it.

https://github.com/checkstyle/checkstyle/blob/4b0c63277dd4554e101f6428208af66a2ac45590/src/main/java/com/puppycrawl/tools/checkstyle/api/JavadocCommentsTokenTypes.java#L49-L72


```
Content of Test.java:
     1	class Test {
     2	
     3	    /**
     4	     * This method uses {@systemProperty user.home} system property.
     5	     */
     6	    void method() {
     7	
     8	    }
     9	}

Executing: java -jar ./checkstyle-12.3.0-all.jar -J Test.java
COMPILATION_UNIT -> COMPILATION_UNIT [1:0]
`--CLASS_DEF -> CLASS_DEF [1:0]
    |--MODIFIERS -> MODIFIERS [1:0]
    |--LITERAL_CLASS -> class [1:0]
    |--IDENT -> Test [1:6]
    `--OBJBLOCK -> OBJBLOCK [1:11]
        |--LCURLY -> { [1:11]
        |--METHOD_DEF -> METHOD_DEF [6:4]
        |   |--MODIFIERS -> MODIFIERS [6:4]
        |   |--TYPE -> TYPE [6:4]
        |   |   |--BLOCK_COMMENT_BEGIN -> /* [3:4]
        |   |   |   |--COMMENT_CONTENT -> *\n     * This method uses {@systemProperty user.home} system property.\n      [3:6]
        |   |   |   |   `--JAVADOC_CONTENT -> JAVADOC_CONTENT [3:7]
        |   |   |   |       |--NEWLINE -> \n [3:7]
        |   |   |   |       |--LEADING_ASTERISK ->      * [4:0]
        |   |   |   |       |--TEXT ->  This method uses  [4:6]
        |   |   |   |       |--JAVADOC_INLINE_TAG -> JAVADOC_INLINE_TAG [4:24]
        |   |   |   |       |   `--SYSTEM_PROPERTY_INLINE_TAG -> SYSTEM_PROPERTY_INLINE_TAG [4:24]
        |   |   |   |       |       |--JAVADOC_INLINE_TAG_START -> {@ [4:24]
        |   |   |   |       |       |--TAG_NAME -> systemProperty [4:26]
        |   |   |   |       |       |--TEXT ->   [4:40]
        |   |   |   |       |       |--IDENTIFIER -> user.home [4:41]
        |   |   |   |       |       `--JAVADOC_INLINE_TAG_END -> } [4:50]
        |   |   |   |       |--TEXT ->  system property. [4:51]
        |   |   |   |       |--NEWLINE -> \n [4:68]
        |   |   |   |       `--TEXT ->       [5:0]
        |   |   |   `--BLOCK_COMMENT_END -> */ [5:5]
        |   |   `--LITERAL_VOID -> void [6:4]
        |   |--IDENT -> method [6:9]
        |   |--LPAREN -> ( [6:15]
        |   |--PARAMETERS -> PARAMETERS [6:16]
        |   |--RPAREN -> ) [6:16]
        |   `--SLIST -> { [6:18]
        |       `--RCURLY -> } [8:4]
        `--RCURLY -> } [9:0]
```

what we try to archive is validation on BLOCK_COMMENT_END to be same line as BLOCK_COMMENT_BEGIN or alone   on line.

> Therefore, introducing a new property only for this case is a good idea.

Unfortunately , I disagree, it will be complicated for user to understand and will be hard to maintain such logic, we more and more leaning away from big Checks, and try to focus on simple operation but it very well and easy to be extended for customization if users request. 
We need new Check.

We can make BlockCommentPosition Check with property that allow "alone_or_singleline" policy.
https://checkstyle.sourceforge.io/property_types.html#RightCurlyOption , it will be similar to `alone_or_singleline` and as you can see in future we can do more options if users be interested. But I think the only reasonable other value is "alone", if some users do not like single line javadocs.

**Comment by Zopsss** [comment]
> I do not recommend to ad support for alignment of third position from the beginning, but as we have new Check , extension of it will be easy.
>
>Lets create new issue and start discussion of new Check. Ones implementation is done, we can comeback to this issue and activate it google style config.

When we were discussing about implementing the leading asterisk align check, we had a discussion about these alignments: https://github.com/checkstyle/checkstyle/issues/6723#issuecomment-493777285

At that time, we decided not to implement these alignments, we decided if any user demands for it only then we should implement it by extending this check. So we don't need a new check just for this. We can extend this check and introduce a new property, which will allow users to specify the alignments of leading asterisks, like: CENTER, LEFT, RIGHT.

**Comment by Zopsss** [comment]
> > Therefore, introducing a new property only for this case is a good idea.
> 
> Unfortunately , I disagree, it will be complicated for user to understand and will be hard to maintain such logic, we more and more leaning away from big Checks, and try to focus on simple operation but it very well and easy to be extended for customization if users request. We need new Check.

Thank you for the detailed explanation. I agree with you, to check if `/**` & `*/` are alone one the line, we can create a new check. 

> We can make BlockCommentPosition Check with property that allow "alone_or_singleline" policy. https://checkstyle.sourceforge.io/property_types.html#RightCurlyOption , it will be similar to `alone_or_singleline` and as you can see in future we can do more options if users be interested. But I think the only reasonable other value is "alone", if some users do not like single line javadocs.

We already have [BlockCommentPosition.java](https://github.com/checkstyle/checkstyle/blob/master/src/main/java/com/puppycrawl/tools/checkstyle/utils/BlockCommentPosition.java), it is utility class to check for javadoc positions. 

We need some other name, this new check should have following properties:

1. singleline_javadoc_allowed: if true, single line javadocs will be allowed, if false then it will give violation for it.
2. alone_or_singleline: if true, `/**` & `*/` should always be alone on the line, if false then they can be preceded and followed by other code, for example:
```java
int x = 99; /** hello there
* how are you?
*/ int a = 45;
```

**Comment by Praveen7294** [comment]
>Unfortunately , I disagree, it will be complicated for user to understand and will be hard to maintain such logic, we more and more leaning away from big Checks, and try to focus on simple operation but it very well and easy to be extended for customization if users request.
We need new Check.

ok, i agree with you we should introduce a new check.

>We need some other name, this new check should have following properties:

>1. singleline_javadoc_allowed: if true, single line javadocs will be allowed, if false then it will give violation for it.
>2. alone_or_singleline: if true, /** & */ should always be alone on the line, if false then they can be preceded and followed by other code, for example:
>```
>int x = 99; /** hello there
>* how are you?
>*/ int a = 45;
>```

Since we are trying to achieve validation that `BLOCK_COMMENT_END` is either on the same line as `BLOCK_COMMENT_BEGIN` or alone on its own line, the check name could be `BlockCommentEndPosition` with below properties:

1. allowSameLine - true (default)
    - `true` -> `*/` must be on the same line as `/*`
    - `false` ->  same line is optional / allowed if other rules allow
2. allowOnOwnLine - true (default)
    - true -> `*/` must be alone on its own line
    - false -> own line is optional / allowed if other rules allow

The posibility of `*/` position are:
case 1:
```
/**        */
```
case 2:
```
/**
 *
 */
```
case 3:
```
/**
 * comment */
```
### case study
| Scenario | strictlyOnSameLine: true, strictlyOnOwnLine: false | strictlyOnSameLine: false, strictlyOnOwnLine: true | strictlyOnSameLine: true, strictlyOnOwnLine: true | strictlyOnSameLine: false, strictlyOnOwnLine: false |
|----------|-------------------------------------------|--------------------------------------------|-------------------------------------------|--------------------------------------------|
| case 1    | no violation   | violation         | no violation    | no violation    |
| case 2    | violation        | no violation    | no violation    | no violation    |
| case 3    | violation        | violation         | violation         | no violation    |

is this proposal good for a new check?

**Comment by romani** [comment]
>below properties:

Having 2 properties will be complicated, as you need to make all combinations of their values.
If we make single property, with values alone_or_singleline and alone. We make user to define whole strategy for validation. Property name can be "strategy".

**Comment by Praveen7294** [comment]
> Having 2 properties will be complicated, as you need to make all combinations of their values.

Yes but it covers all posibility of position of `*/`. And if any user do not want restriction on position of `*/`, it will helpful to user.

>If we make single property, with values alone_or_singleline and alone. We make user to define whole strategy for validation. Property name can be "strategy".

if we use single property with values `alone` and `alone_or_sinleline`, we might not able to cover case 3

- `alone` : allow only own line and report violation when `*/` not in it's own line.
- `alone_or_singleline` : allow both own line and singleline comment.

Strategy | alone | alone_or_singleline | 
-- | -- | -- 
case 1 | violation | ok 
case 2 | ok | ok 
case 3 | violation | violation

**Comment by Praveen7294** [comment]
> if any user do not want restriction on position of `*/`.

Now, i understand if any user do not want restriction on position of `*/` then user should avoid `BlockCommentEndPosition` check.

i am creating new issue for `BlockCommentEndPosition` check.

we introduce `strategy` property with values `alone` and `alone_or_sinleline`:

- `alone` : allow only own line and report violation when `*/` not in it's own line.
- `alone_or_singleline` : allow both own line and singleline comment.

Strategy | alone | alone_or_singleline
-- | -- | --
case 1 | violation | ok
case 2 | ok | ok
case 3 | violation | violation

</markdown-accessiblity-table>
</body>
</html>

**Comment by Zopsss** [comment]
To follow the style guide properly, we need to ensure 2 things:

1. Single line javadocs are allowed:

```java
/** Hello there. */
```

2. If the javadoc cannot be fit in one line then it should be of multiple lines, and the content cannot be preceded by `/**` and followed by `*/`, i.e the `/**` & `*/` should always be alone on the line.

```java
/** Hello there
* How are you? */
```

The above example should give violation. 


The new check should have 1 property, called `allow_single_line_comments`, which will default to `true`, if false then single line comments will be not allowed. ( Note that this will be a general purpose check, so other users who doesn't use our google style guide implementation can also use this check however they want. )

The check's main purpose is to ensure that `/**`& `*/` should be alone on the line, if `allow_single_line_comments` is set to true, only then they're allowed on single line, but even if that property is enabled, it should log violation for the above mentioned incorrect example. 

The check should be focused on Comments only. It should be easy to implement. 


---

I have one question though, do we handle following type of incorrect javadoc placements?

```java
int x = 99; /** Hey there
* How are you */

int a = 45; /** Hey there. */

/** Hey there.
* How are you? */ int p = 11;
```

Such types of comments are incorrectly placed, @Praveen7294 can you check if we already log violation for the above mentioned cases?

**Comment by Praveen7294** [comment]
@Zopsss 

>I have one question though, do we handle following type of incorrect javadoc placements?
>
>```
>int x = 99; /** Hey there
>* How are you */
>
>int a = 45; /** Hey there. */
>
>/** Hey there.
>* How are you? */ int p = 11;
>```
>Such types of comments are incorrectly placed, @Praveen7294 can you check if we already log violation for the above mentioned cases?

### Test.java

```
/**                                                          // 1
 * Testing.                                                  // 2
 */                                                          // 3
public class Test {                                          // 4
                                                             // 5
  /**                                                        // 6
   * Test Method.                                            // 7
   */                                                        // 8
  public void foo() {                                        // 9
                                                             // 10
    int x = 99; /** Hey there                                // 11
    * How are you */                                         // 12
                                                             // 13
    int a = 45; /** Hey there. */                            // 14
                                                             // 15
    /** Hey there.                                           // 16
    * How are you? */ int p = 11;                            // 17
                                                             // 18
  }                                                          // 19
}                                                            // 20
```

### config
https://github.com/checkstyle/checkstyle/blob/master/src/main/resources/google_checks.xml

### output
```
$ java -jar checkstyle-13.0.0-all.jar -c google_checks.xml Test.java
Starting audit...
[WARN] F:\GitHub\headhtmltagname\Test.java:11:17: Javadoc comment is placed in the wrong location. [InvalidJavadocPosition]
[WARN] F:\GitHub\headhtmltagname\Test.java:14:17: Javadoc comment is placed in the wrong location. [InvalidJavadocPosition]
[WARN] F:\GitHub\headhtmltagname\Test.java:16:5: Javadoc comment is placed in the wrong location. [InvalidJavadocPosition]
Audit done.
```

**Comment by Praveen7294** [comment]
And also

- Indentation of `/**` (Opening javadoc tag) is checked by `CommentsIndentation`.
- Alignment of `*` (leading asterisk) is checked by `JavadocLeadingAsteriskAlign` with respect to `/**`. Therefore, if the position of `*` is incorrect, a violation will also be reported for the leading asterisk alignment.
- Alignment of `*/` is checked by `JavadocLeadingAsteriskAlign` when it has it's own line.

The only case not checked by any existing check is when */ does not have its own line. see https://github.com/checkstyle/checkstyle/issues/18273#issuecomment-3662887751 i already mentioned.

so the new check should only check the position of `*/`. There are three possibility of position of `*/` :
#### case 1:
```
/**        */
```
#### case 2:
```
/**
 *
 */
```
#### case 3:
```
/**
 * comment */
```

**case 1** and **case 2** is allowed by the google style guide.

---

> The new check should have 1 property, called `allow_single_line_comments`, which will default to true.

@Zopsss can you clarify what should happen when `allow_single_line_comments` is true.

please note that google style guide allow single line comment and mutli line comment when comments not fit into a single line.

**Comment by Zopsss** [comment]
> can you clarify what should happen when allow_single_line_comments is true.


When this property is set to true, we'll allow `/**` & `/*` to be on the same line, i.e following case will be allowed:

```java
/** Hey there. */
```

For more clarity, even if this property is set to true, following examples would still log violation:

```java
/** 
  * Hey there. */ // Violation, ending tag should be alone on line

/** Hey there. // Violation, starting tag should be alone on line
  * How are you? 
  */ 
```

PS: violation messages should be more precise.

**Comment by Praveen7294** [comment]
@Zopsss Having `allow_single_line_comment` property doesn't obey the google style guide.

because below cases is allowed by google style guide
```
/** 
 * Hey there.
 * How are you? 
 */ 
```

So, `strategy` property is good with value `alone` or `alone_or_singleline` in new check. For explanation see https://github.com/checkstyle/checkstyle/issues/18273#issuecomment-3727777796

**Comment by Zopsss** [comment]
Sorry I'm not clearly able to understand the comment you pointed out. 

The case you mentioned is allowed by the style guide and the check would also not produce violation for it. 

My suggested check would only give violation when `/**` or `*/` are not alone on the line. Please feel free to ask any more questions if you have.

**Comment by Praveen7294** [comment]
According to your suggestion, i understand

- if `allow_single_line_comment` is true, the following case is allowed:
   ```
   /** Hey there. */
   ```
   and give violation on below cases:
   ```
   /** 
    * Hey there. */ // Violation, ending tag should be alone on line

   /** Hey there. // Violation, starting tag should be alone on line
    * How are you? 
    */
   ```

- if `allow_single_line_comment` is false, the following cases is allowed:
   ```
   /** 
    * Hey there. */ // Violation, ending tag should be alone on line

   /** Hey there. // Violation, starting tag should be alone on line
    * How are you? 
    */
   ```
   and give violation on below case:
   ```
   /** Hey there. */
   ```

please correct me if i am wrong.

According to google style guide, below both cases are valid:
```
/** Hey there. */

/** Hey there. // Violation, starting tag should be alone on line
 * How are you? 
 */
```
but your suggestion allowed any one valid even `allow_single_line_comment` is true or false.

**Comment by Zopsss** [comment]
The `allow_single_line_comment` property will allow to have single line comments, it doesn't matter if it is set to true or false, the following cases will always give violation:

```java
/** Hey there.
  * How are you?
  */

/**
  * Hey there. */
```

`allow_single_line_comment`'s job is to only allow for single line comments, that's it. All the other invalid cases will still stay as invalid.

**Comment by Praveen7294** [comment]
There is one thing if javadoc less than 100 characters then javadoc should be fit in single line. But in most of the cases javadoc is more than 100 characters should be placed in multiline javadoc.

>```
>/** Hey there.
>  * How are you?
>  */
>```

In above case, this javadoc is less than 100 characters and it is placed in multiline javadoc then there is a violation.

Is this is ur intuition?

**Comment by Zopsss** [comment]
It depends on user if he wants his javadoc to be in multiple line or in single line. If javadoc is less than 100 chars then he should be able to put it in single line, if the line exceeds that character limit then [LineLength](https://checkstyle.sourceforge.io/checks/sizes/linelength.html) will give violation for it.

The case you mentioned will give violation because it is a multi line comment, the javadoc content is starting immediately after `/**` which is not allowed by google.

If it was a single line comment like:

```
/** Hey there, how are you? */
```

then it will not give violation, but since it is a multi line comment so it will give violation.

**Comment by Praveen7294** [comment]
> The case you mentioned will give violation because it is a multi line comment, the javadoc content is starting immediately after /** which is not allowed by google.

Yes i know, this is detected by [JavadocContentLocation](https://checkstyle.org/checks/javadoc/javadoccontentlocation.html#JavadocContentLocation). 

Sorry, I did not give a proper example earlier. Below is the correct example:

```
/** 
 * Hey there, Testing testing testing testing testing testing testing testing testing testing testing testing.
 * This Javadoc is taken as example.
 */
```

above example should be allowed right? (multiline javadoc because cannot fit into single line if we try then line exceed characters limit.)

> It depends on user if he wants his javadoc to be in multiple line or in single line.

Yes i want same behavior from our new checks. So i took `strategy` property with value `alone` or `alone_or_singleline`. Something like this https://checkstyle.sourceforge.io/property_types.html#RightCurlyOption. we are using this Right curly (`}`)and here we will use for closing asterisk (`*/`).

- **alone** : `*/` should be on own line. i.e only multiline comment allowed.

   **Valid case**:
   ```
   /** 
    * Hey there, Testing testing testing testing testing testing testing testing testing testing testing testing.
    * This Javadoc is taken as example.
    */
   ```

   **Invalid cases** (due to `*/` is not on its own line):
   ```
   /** Hey there, how are you? */


   /** 
    * Hey there, how are you? */
   ```
   [New open jdk](https://cr.openjdk.org/~alundblad/styleguide/index-v6.html#toc-commenting-code) chooses `//` for singleline comment and `/* comment */` for multiline comment. `alone` value is fit for this style guide.

- **alone_or_singleline**: `*/` should be allowed for singleline comment and multiline comment.

   **Valid case**:
   ```
   /** Hey there, how are you? */


   /** 
    * Hey there, Testing testing testing testing testing testing testing testing testing testing testing testing.
    * This Javadoc is taken as example.
    */
   ```

   **Invalid cases**:
   ```
   /** 
    * Hey there, how are you? */
   ```

Most of the style guide allowed singleline comment and multiline comment ([google style](https://google.github.io/styleguide/javaguide.html#s7.1.1-javadoc-multi-line), [sun code](https://checkstyle.org/styleguides/sun-code-conventions-19990420/CodeConventions.doc4.html#a385), ) .so this is the reason for taking `strategy` property with value `alone_or_singleline`.

@Zopsss please tell me if you have other strategy to implement our new check.

**Comment by Zopsss** [comment]
> > The case you mentioned will give violation because it is a multi line comment, the javadoc content is starting immediately after /** which is not allowed by google.
> 
> Yes i know, this is detected by [JavadocContentLocation](https://checkstyle.org/checks/javadoc/javadoccontentlocation.html#JavadocContentLocation).

We can omit this check if we go with my approach of the Check. And this check is only for Javadoc comments, so I don't think it will be able to detect `/* .... */` comments for [New Open JDK standards](https://cr.openjdk.org/~alundblad/styleguide/index-v6.html#toc-commenting-code).

---

> ```
> /** 
>  * Hey there, Testing testing testing testing testing testing testing testing testing testing testing testing.
>  * This Javadoc is taken as example.
>  */
> ```
> 
> above example should be allowed right? (multiline javadoc because cannot fit into single line if we try then line exceed characters limit.)

Yes it is allowed.

---

> please tell me if you have other strategy to implement our new check.

Your proposal of the new check is good but it has some gaps, specifically with respect to:

> > Yes i know, this is detected by [JavadocContentLocation](https://checkstyle.org/checks/javadoc/javadoccontentlocation.html#JavadocContentLocation).
> 
> We can omit this check if we go with my approach of the Check. And this check is only for Javadoc comments, so I don't think it will be able to detect `/* .... */` comments for [New Open JDK standards](https://cr.openjdk.org/~alundblad/styleguide/index-v6.html#toc-commenting-code).

The JavadocContentLocation is only useful for Javadoc comments. 

With my approach, I think the check should be able to detect both types of comments: `/** ... */`, and `/* .... */` because of the AST structure:


### Example:

```java
/**
 * Test
 */
public class Test {
    /*
     * test
     */
    int x;
}
```


### AST:

```
$ java -jar .\checkstyle-10.25.0-all.jar -J .\Test.java
COMPILATION_UNIT -> COMPILATION_UNIT [4:0]
`--CLASS_DEF -> CLASS_DEF [4:0]
    |--MODIFIERS -> MODIFIERS [4:0]
    |   |--BLOCK_COMMENT_BEGIN -> /* [1:0]
    |   |   |--COMMENT_CONTENT -> *\r\n * Test\r\n  [1:2]
    |   |   |   `--JAVADOC -> JAVADOC [1:3]
    |   |   |       |--NEWLINE -> \r\n [1:3]
    |   |   |       |--LEADING_ASTERISK ->  * [2:0]
    |   |   |       |--TEXT ->  Test [2:2]
    |   |   |       |--NEWLINE -> \r\n [2:7]
    |   |   |       |--TEXT ->   [3:0]
    |   |   |       `--EOF -> <EOF> [3:1]
    |   |   `--BLOCK_COMMENT_END -> */ [3:1]
    |   `--LITERAL_PUBLIC -> public [4:0]
    |--LITERAL_CLASS -> class [4:7]
    |--IDENT -> Test [4:13]
    `--OBJBLOCK -> OBJBLOCK [4:18]
        |--LCURLY -> { [4:18]
        |--VARIABLE_DEF -> VARIABLE_DEF [8:4]
        |   |--MODIFIERS -> MODIFIERS [8:4]
        |   |--TYPE -> TYPE [8:4]
        |   |   |--BLOCK_COMMENT_BEGIN -> /* [5:4]
        |   |   |   |--COMMENT_CONTENT -> \r\n     * test\r\n      [5:6]
        |   |   |   `--BLOCK_COMMENT_END -> */ [7:5]
        |   |   `--LITERAL_INT -> int [8:4]
        |   |--IDENT -> x [8:8]
        |   `--SEMI -> ; [8:9]
        `--RCURLY -> } [9:0]
```

The `BLOCK_COMMENT_BEGIN` and `BLOCK_COMMENT_END` are common Javadoc and block comments.

**Comment by Praveen7294** [comment]
> so I don't think it will be able to detect `/* .... */` comments for [New Open JDK standards](https://cr.openjdk.org/~alundblad/styleguide/index-v6.html#toc-commenting-code).

Yes i checked, it is not detected.

```
/** Testing.
 * 
 */
public class Test {

  /** 
   * Test.
   */
  public void test() {
    
    /* This is a multiline comment testing.
     * 
     */
  }
}
```
Cli:
```
$ java -jar checkstyle-13.0.0-all.jar -c google_checks.xml Test.java
Starting audit...
[WARN] F:\GitHub\headhtmltagname\Test.java:1:1: Javadoc content should start from the next line after /**. [JavadocContentLocation]
Audit done.
```

---
> The JavadocContentLocation is only useful for Javadoc comments.
>
> With my approach, I think the check should be able to detect both types of comments: `/** ... */`, and `/* .... */`. 

But there is no formatting rule defined for multiline comment `/* .... */` only mentioned that `For single line comments, use end-of-line comments (//) otherwise use multiline comments (/* … */).` in [New Open JDK standards](https://cr.openjdk.org/~alundblad/styleguide/index-v6.html#toc-commenting-code). Do we need to include this case in new checks.

And [New Open jdk javadoc](https://cr.openjdk.org/~alundblad/styleguide/index-v6.html#toc-javadoc) have formatting rule. i think we should implement new check for javadoc.

@Zopsss what's your thought do we need to include multiline comment in new check?

**Comment by Zopsss** [comment]
I roughly checked New Open JDK Javadoc rule, it seems similar to the google style guide. 

With my approach of the check, we can omit JavadocContentLocation, it might also be useful for `/* .... */` comments as well while your check if more specifically for `*/`


@romani Wdyt about this? Which approach should we take? References of the comments which explains my approach: https://github.com/checkstyle/checkstyle/issues/18273#issuecomment-3744294905, https://github.com/checkstyle/checkstyle/issues/18273#issuecomment-3770930688

@Praveen7294's approach: https://github.com/checkstyle/checkstyle/issues/18273#issuecomment-3768454184

**Comment by Praveen7294** [comment]
> while your check if more specifically for `*/`.

Yes, my check is specifically for `*/`, because if we strictly enforce that `/**` must always be on its own line, then we are implicitly restricting Javadoc content to start only from the second line of the Javadoc comment. Due to this, the new check reports a violation for this case: https://checkstyle.org/checks/javadoc/javadoccontentlocation.html#Example2-config

But if your plan is to omit [JavadocContentLocation](https://checkstyle.org/checks/javadoc/javadoccontentlocation.html#JavadocContentLocation), then I can also enforce that `/**` must always be on its own line.

**Comment by Zopsss** [comment]
IMO we should go with my approach, the new check will be for both: block comments and javadoc comments. It can be used by people who uses our google style guide implementation as well as others who doesn't use it. The check will target both: starting tag and ending tag of the comments, it will be more easy to understand, and use and i guess implement as well.
